File: frontend/src/services/notification.service.ts
import { api } from "@/lib/api-client";
import { ENDPOINTS } from "@/constants/api";
import { NotificationResponse } from "@/types/notification";

interface PageResponse<T> {
  content: T[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  last: boolean;
}

export const notificationService = {
  getNotifications: async (
    page = 0,
    size = 10,
    readStatus: "ALL" | "UNREAD" = "ALL"
  ): Promise<PageResponse<NotificationResponse>> => {
    return api.get<PageResponse<NotificationResponse>>(ENDPOINTS.NOTIFICATIONS.LIST, {
      params: { page, size, readStatus },
    });
  },

  getUnreadCount: async (): Promise<number> => {
    return api.get<number>(ENDPOINTS.NOTIFICATIONS.UNREAD_COUNT);
  },

  markAsRead: async (id: string): Promise<void> => {
    await api.put(ENDPOINTS.NOTIFICATIONS.READ(id));
  },

  markAllAsRead: async (): Promise<void> => {
    await api.put(ENDPOINTS.NOTIFICATIONS.READ_ALL);
  },

  deleteNotification: async (id: string): Promise<void> => {
    await api.delete(ENDPOINTS.NOTIFICATIONS.DELETE(id));
  },

  deleteAllNotifications: async (): Promise<void> => {
    await api.delete(ENDPOINTS.NOTIFICATIONS.DELETE_ALL);
  },
};

File: backend/src/main/java/com/vibecart/api/modules/notification/service/NotificationService.java
package com.vibecart.api.modules.notification.service;

import com.vibecart.api.common.dto.PageResponse;
import com.vibecart.api.modules.notification.dto.event.InAppNotificationEvent;
import com.vibecart.api.modules.notification.dto.request.UpdatePreferencesRequest;
import com.vibecart.api.modules.notification.dto.response.NotificationResponse;
import com.vibecart.api.modules.notification.dto.response.PreferencesResponse;

public interface NotificationService {
    PageResponse<NotificationResponse> getNotifications(String username, int page, int size, String readStatus);
    long getUnreadCount(String username);
    void markAsRead(String notificationId, String username);
    void markAllAsRead(String username);
    void deleteNotification(String notificationId, String username);
    void deleteAllNotifications(String username);
    void processAndBroadcast(InAppNotificationEvent event);

    PreferencesResponse getPreferences(String username);
    PreferencesResponse updatePreferences(String username, UpdatePreferencesRequest request);
}
